apps/web/src/views/Swap/V3Swap/hooks/useWallchain.ts

```diff
@@ -0,0 +1,235 @@
+}
+const wrappedLoadData = limiter.wrap(loadData)
+
+const extractAddressFromCurrency = (currency: Currency): `0x${string}` => {
+  return currency.isNative ? '0xeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee' : (currency as Token).address
+}
+
+const extractTokensFromTrade = (trade: SmartRouterTrade<TradeType> | undefined | null) => {
+  const inputCurrency = trade?.inputAmount?.currency
+  const outputCurrency = trade?.outputAmount?.currency
+  const srcToken = inputCurrency ? extractAddressFromCurrency(inputCurrency) : false
+  const dstToken = outputCurrency ? extractAddressFromCurrency(outputCurrency) : false
+
+  return [srcToken, dstToken] as [false | `0x${string}`, false | `0x${string}`]
+}
+
+function useWallchainSDK() {
+  const { data: walletClient } = useWalletClient()
+  const { chainId } = useActiveChainId()
+  const { data: wallchainSDK } = useSWRImmutable(
+    chainId === ChainId.BSC && walletClient && ['wallchainSDK', walletClient.account, walletClient.chain],
+    async () => {
+      const WallchainSDK = (await import('@wallchain/sdk')).default
+      return new WallchainSDK({
+        keys: WallchainKeys,
+        provider: walletClient?.transport as TOptions['provider'],
+        overrideAddresses,
+      })
+    },
+  )
+
+  return wallchainSDK
+}
+
+const wallchainStatusAtom = atom<WallchainStatus>('pending')
+export function useWallchainStatus() {
+  return useAtom(wallchainStatusAtom)
+}
+
+export function useWallchainApi(
+  trade?: SmartRouterTrade<TradeType>,
+  deadline?: bigint,
+  feeOptions?: FeeOptions,
+): [WallchainStatus, string | undefined, [TMEVFoundResponse['searcherRequest'], string] | undefined] {
+  const [approvalAddress, setApprovalAddress] = useState<undefined | string>(undefined)
+  const [status, setStatus] = useWallchainStatus()
+  const [masterInput, setMasterInput] = useState<undefined | [TMEVFoundResponse['searcherRequest'], string]>(undefined)
+  const { data: walletClient } = useWalletClient()
+  const { account } = useAccountActiveChain()
+  const [allowedSlippageRaw] = useUserSlippage() || [INITIAL_ALLOWED_SLIPPAGE]
+  const allowedSlippage = useMemo(() => basisPointsToPercent(allowedSlippageRaw), [allowedSlippageRaw])
+
+  const sdk = useWallchainSDK()
+
+  const swapCalls = useSwapCallArguments(trade, allowedSlippage, account, deadline, feeOptions)
+
+  useEffect(() => {
+    if (!sdk || !walletClient || !trade) return
+    if (trade.routes.length === 0 || trade.inputAmount.currency.chainId !== ChainId.BSC) return
+    const includesPair = trade.routes.some(
+      (route) =>
+        (route.inputAmount.wrapped.currency.equals(WallchainPairs[0]) &&
+          route.outputAmount.wrapped.currency.equals(WallchainPairs[1])) ||
+        (route.inputAmount.wrapped.currency.equals(WallchainPairs[1]) &&
+          route.outputAmount.wrapped.currency.equals(WallchainPairs[0])),
+    )
+    if (includesPair) {
+      setStatus('pending')
+      wrappedLoadData(account, sdk, swapCalls)
```

Comment:
For data loading we want to ensure that response always matches the swapCalls. It means, it would be better to drop task with bottleneck (false negative), then end up with cached value than doesn't match current route (false positive). Though, api is called immediately during a swap call, in case of false positive we would have wrong approval address being checked.